Speed up `npm ci` by caching `node_modules`

kevin940726 opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Speed up the CI pipeline by caching `node_modules`.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Faster CI === Happier developers? ❤️ 
Slightly related to #33532, #44679, and other similar performance related issues.

The official `setup-node` action by default only caches the global cache (`~/.npm` in *nix). However, it's still taking roughly 2 minutes to install all the packages in **every** workflow. By also caching `node_modules`, we can dramatically cut down the installation time by 75%.

Note that in the official document, it's [not recommended](https://github.com/actions/cache/blob/main/examples.md#node---npm) to cache `node_modules`.

> Note: It is not recommended to cache `node_modules`, as it can break across Node versions and won't work with `npm ci`

However,
1. We rarely change our Node versions.
2. It won't work with `npm ci` because `npm ci` will delete `node_modules` before installing the modules. We can simply avoid running `npm ci` if there's a cache hit.

TBH, I'm not really sure why it's not more of an established pattern. Probably someone knows better?

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
I'm abstracting the steps into a [composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) called `setup-node`. It roughly does the following:

1. Setup Node.js to the desired version.
3. Cache global npm cache (`~/.npm`) to speed up download time.
4. Cache `node_modules` to speed up installation time.
5. If the cache missed in step 3, run `npm ci --ignore-scripts` to download the modules.
6. Rebuild the dependencies (`npm rebuild`) if necessary.
7. Run `npm run postinstall`.

Then, we just have to replace every `actions/setup-node` + `npm ci` duo with this composite action `setup-node` in each workflow. Note that this action doesn't help if the workflow doesn't need to run `npm ci` though. What's good is that such workflows can continue using the original `action/setup-node` since they share the same global npm cache.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

The result can be verified by comparing the installation time in the workflows.

An example [run](https://github.com/WordPress/gutenberg/actions/runs/3513636215/jobs/5887600610) in this PR [took `1m 57s`](https://github.com/WordPress/gutenberg/actions/runs/3513636215/jobs/5886694027) to run the composite action when the cache missed. Re-running the job again, it then only [took `20s`](https://github.com/WordPress/gutenberg/actions/runs/3513636215/jobs/5887600610) to run the same composite action when the cache hit.

Compare that to [`trunk`](https://github.com/WordPress/gutenberg/actions/runs/3515221317/jobs/5890206097). The equivalent actions duo (`actions/setup-node` + `npm ci`) took [`2m 5s`](https://github.com/WordPress/gutenberg/actions/runs/3515221317/jobs/5890206097) to run when the cache hit. We can draw this table below with limited source data:

Step duration in seconds | Cache miss | Cache hit
--- | --- | ---
Caching only global npm cache (`trunk`) | [`138s`](https://github.com/WordPress/gutenberg/actions/runs/3477923708/jobs/5814666495) | [`125s`](https://github.com/WordPress/gutenberg/actions/runs/3515221317/jobs/5890206097)
Also caching `node_modules` (this PR) | [`117s`](https://github.com/WordPress/gutenberg/actions/runs/3513636215/jobs/5886694027)[^1] | [`20s`](https://github.com/WordPress/gutenberg/actions/runs/3513636215/jobs/5887600610)

[^1]: There's a post action below that took `1m 29s`. It's likely a race condition issue when multiple jobs were trying to save the same cache.

Even though sometimes the duration varies a lot between runs, the performance improvement is still undeniably outstanding.

## Trade-offs

The improvement is obviously not free, there's always a cost. In this case, the main cost is the cache size. The original global npm cache has a size of (roughly) `270MB` per key. On top of that, this PR adds an additional `270MB` of `node_modules` cache per key. Because `node_modules` won't work across different OSes or Node versions, we will have to also cache them for `Windows`, `macOS`, etc. They are adding up and will soon exceed the `10GB` threshold of the [maximum cache size in a repository](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy).

At the time of writing, we currently have [77 caches](https://github.com/WordPress/gutenberg/actions/caches) (counting these `node_modules` cache) and the least used cache is only 3 days old. If we're not careful, we could cause cache thrashing and defeat the whole purpose of caching.

One possible solution to this is to avoid caching `node_modules` or fallback to a common cache in some of the less-used situations. For instance, we could try grouping all OSes in the same key, but require Windows and macOS to always re-run `npm rebuild` after restoring the cache. The same could also apply to different Node versions as well. This technique is potentially unsafe though, and might cause confusion to contributors if the cache fails.

With that being said, I think the additional total of 1GB of cache is still worth the time save. We can monitor the impact of these caches for a while to be sure.

## Screenshots or screencast <!-- if applicable -->
![image](https://user-images.githubusercontent.com/7753001/203087545-0e24fd3c-c004-4b6d-9167-9c9e8eecf958.png)

kevin940726 committed:
Try custom setup-node composite action for aggressive caching

kevin940726 committed:
Add comment about GHA bug

kevin940726 commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,58 @@
+name: 'Setup Node'
```
I'm not entirely happy with the name. Maybe we should make it explicit that it also handles installing dependencies?

desrosj replied:
Hmm, how about something like "Configure Node.js and install npm dependencies while managing all aspects of caching"?

It looks like this `name` is not displayed anywhere. For example, in [this workflow associated with this PR](https://github.com/WordPress/gutenberg/actions/runs/4111626373/jobs/7095591254#step:3:1), the text displayed for the step calling this composite action is the one defined in that respective workflow. I'm not sure where this will display, actually. May be better to give a full short description here.

kevin940726 replied:
I changed the name a bit and moved your suggestion to the `description` field.

kevin940726 commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,58 @@
+          uses: actions/cache@v3
+          with:
+              path: '**/node_modules'
+              key: node_modules-${{ runner.os }}-${{ steps.node-version.outputs.NODE_VERSION }}-${{ hashFiles('package-lock.json') }}
```
[`restore-keys`](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#matching-a-cache-key) is not needed because we'll be running `npm ci` if there isn't an exact match anyway. `npm ci` will remove all of the `node_modules`.

dmsnell replied:
do we need to include the `npm` version in here? I know we already have issues where newer `npm` won't work because of conflicts with package resolution between the versions, or in the `package-lock.json` format.

kevin940726 replied:
I don't think so. If a newer `npm` version invalidates the cache then it should also change `package-lock.json`. A common example is the `lockfileVersion` field in the `package-lock.json` file.

dmsnell replied:
right, but if we update the `npm` version and use the cache from the old one then won't that lead to a mismatch in the installed packages?

kevin940726 replied:
I don't think an updated npm version will install a different package, will it? Even if it will, the `package-lock.json` file will change, and the commit that changes the file will invalidate the cache.

dmsnell replied:
I'm not sure. I don't know if anything will change except for the built dependencies. It seems like if I compile the existing project with different versions of `npm` I _have_ to clear out `node_modules` or `npm ci` fails, and possibly won't run with a mix of modules built in the old version and some built in the new. I'm just double-checking here because this introduces what could be a opportunity for things to break on account of this, and again, in a very nebulous manner that won't appear obvious in the logs.

you are probably right; but I'd want us to be very sure of that

kevin940726 replied:
I don't mind including it in the cache key as it also rarely changes. Appreciate the effort for trying to keep it safe!

kevin940726 commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,58 @@
+          # Comparing against strings is necessary because of a known bug in composite actions.
+          # See https://github.com/actions/runner/issues/1483#issuecomment-1279933184
+          if: ${{ inputs.should-rebuild == 'true' }}
+          run: npm rebuild
```
This is normally not needed because most of the packages don't have `postinstall` scripts. However, some `react-native` related packages do need them so we make this an opt-in.

This is not future-proof as we might include a package that needs `postinstall` in the future. We can run `npx can-i-ignore-scripts` to manually inspect these cases.

dmsnell replied:
can we not run `npm ci` above instead of `npm ci --ignore-scripts`?

also I wonder how much of the impact on performance is related to `--ignore-scripts` - I might measure it and find out.

this seems more of a danger on this change: is there a way we can make it more obvious to someone when their PRs are failing multiple test suites that the reason is that they need to supply `should-rebuild: true`?

I'm not sure how I would expect someone to start from the failure and work back up to this setting.

kevin940726 replied:
Yeah, as I said above, it's not entirely "safe", but I think it's good enough for most cases. We will invalidate the cache whenever OS/Node/`package-lock.json` changes so it's more or less like what we have during local development.

> also I wonder how much of the impact on performance is related to `--ignore-scripts`

Probably not much, because `npm ci` is only run when there isn't a cache hit. Most runs in our measurements have cache hit.

However, `npm rebuild` does contribute to some running time. We use it here to trigger any underlying `postinstall` script in the dependencies that gets ignored by `-ignore-scripts`. Tested with `can-i-ignore-scripts`, most dependencies that have the `postinstall` scripts are `react-native`-related (and local development related), hence I think it makes sense to make it an opt-in when we're building for react-native.

We can try to just use `npm ci` without the `--ignore-scripts` and always run `npm rebuild` though. I think that's unnecessary in most cases and probably will add up 20 to 30 seconds of runtime but it's technically "safer". Caching `node_modules` won't behave exactly like running `npm ci` every time and I think that's okay.

> I'm not sure how I would expect someone to start from the failure and work back up to this setting.

I'd expect it to be pretty visible as it would probably fail the CI and we can look into that. I think it'll only happen when there's a new dependency with `postinstall` get added though, which should be pretty uncommon.

dmsnell replied:
> Most runs in our measurements have cache hit.

where did you measure this? and what was the ratio for `pull_request` events in PRs?

> probably will add up 20 to 30 seconds of runtime but it's technically "safer"

it's usually faster to do the wrong thing 😉 
I'm really nervous about introducing a very subtle behavior that could break things if you don't already happen to know it exists and remember that you need to apply a workaround when it's appropriate.

> it would probably fail the CI and we can look into that

I agree here, but how is it that you would propose someone moves from "my PR failed and I don't know why" to finding this flag?

dmsnell replied:
> I think that's unnecessary in most cases and probably will add up 20 to 30 seconds of runtime

Not that I have investigated why yet, but if I did the test right it doesn't seem to make much difference beyond a couple seconds. There is not enough of an effect with the data collected so far to conclude that there is any measurable impact. I could have done something wrong, or the flag might not be having the impact we expect it to.

Granted, for the unit tests and end-to-end tests there's still that issue of having the multi-modal distribution which could be throwing off the statistics. I haven't investigated the impact that has on the results.

My test branch is #46318 

## Unit Tests workflow
![branch-compare-45932-46318](https://user-images.githubusercontent.com/5431237/206027635-8ad008e2-4dcb-48eb-9ae1-b8971f04b325.png)

## Performance tests workflow
![branch-compare-45932-46318](https://user-images.githubusercontent.com/5431237/206027810-7e483ca5-770c-4ffa-86b7-f7ff82bae09a.png)

kevin940726 replied:
> where did you measure this? and what was the ratio for `pull_request` events in PRs?

I directly looked into the details of each run and see if there's a cache hit. I only looked into this PR's runs as this is where we're measuring.

> it's usually faster to do the wrong thing 😉

It depends on what you think is "wrong". I'm just saying that upon my research this is the trade-off that I'm willing to take. It's obviously subjective and it's totally "right" to think otherwise.

> but how is it that you would propose someone moves from "my PR failed and I don't know why" to finding this flag?

I think we should be able to find this out in a code review. Just like how people work in different OSes than in CI, things are gonna be different and weird things happen in CI. We can also write some guide or log about this if necessary. I agree we should minimize the breaking changes whenever possible but I believe this is something that rarely breaks. However, if the added time of removing `--ignore-scripts` is so little that it doesn't matter then I'll immediately stand on the other side of the fence 😝 .

kevin940726 replied:
I removed `--ignore-scripts` and `npm rebuild` in [1949d5f](https://github.com/WordPress/gutenberg/pull/45932/commits/1949d5f0aa64a7c5a21174d8120b8719a00a1eaa). Seems like it's still working nicely!

I updated the postinstall step to also run the `postinstall` scripts in the packages folder too. This technically doesn't match the behavior of `npm ci` entirely though as it only runs in the `packages` folder but not recursively for all dependencies.

If the `postinstall` scripts only update the `node_modules` folder (which is what most dependencies are doing) then it should be safely cached. However, in `react-native-editor`, the `i18n-cache` files are generated both in `node_modules` and in the `packages/` folder because of linking. During the tests, the `i18n-cache` files are read from the `packages` folder for some reason, hence we need to re-run the `postinstall` script to populate it again. I think this can be seen as an edge case, and possibly there are other ways to solve it too (for example reading from `node_modules` as a fallback).

kevin940726 commented on package.json:
```diff
@@ -193,6 +193,7 @@
 		"fast-glob": "3.2.7",
 		"filenamify": "4.2.0",
 		"glob": "7.1.2",
+		"graceful-fs": "4.2.10",
```
This is to solve an issue that one of the `react-native`-related dependencies is accessing an outdated version of `graceful-fs`.

TBH, I'm not entirely sure why this is only happening in this PR. It might have something to do with the split steps of `npm ci --ignore-scripts` and `npm rebuild`. 🤔

kevin940726 commented:
@dmsnell Pinging you as seems like you have worked on similar things before and might be interested in this 😉 .

dmsnell commented:
@kevin940726 I can monitor this PR and start collecting data on if it has the intended impact.

is two minutes something we think is worth the added complexity of the cache and the predictable cache invalidation problems?

I haven't looked at the other workflows, but there was plenty of performance improvement available in the Performance workflow that didn't involve adding new complexity.

for reference, while working on the performance test suite I was attempting to collect data on at least thirty test runs in the branch to compare against `trunk` and determine if the changes were worth it; not everything I thought would help did.

dmsnell commented:
**Impact on Performance Tests Workflow**

![branch-compare-45932-44907](https://user-images.githubusercontent.com/5431237/205757888-c133456d-0e5c-4281-8f94-9c2c8818325a.png)


 - this branch, 64 samples, µ = 21.02 min
 - `trunk`, 26 samples, µ = 23.76 min
 - the data _strongly suggests this branch is faster for the performance test suite_, with a significance at `P = 0.0001`
 - the measured decrease in average runtime is *2m 44s*

<details><summary>Original comment</summary>
@kevin940726 after around 30 runs of the performance tests here is some initial data; I'm starting to sequence the E2E tests and compare there

![branch-compare-45932-44907](https://user-images.githubusercontent.com/5431237/204050121-5743a5a7-5988-4c01-ab00-988feb017a03.png)

There's not enough evidence after 30 runs to conclude if this makes a difference or not. The results we're seeing are close enough that if we ran this experiment over and over, assuming there were no change in the PR and we weren't caching the `npm` dependencies, then in approximately half of the test runs we'd expect to see a speedup of at least as much as we're seeing with this actual PR. In other words, there's not a definite clear speedup observed in the test runs.

The data does seem like it's a bit more normal than the test runs in trunk, evidenced on the normality graphs. That suggests there could be an element at play whereby this change cuts down on some long-tail skew, but as mentioned above, there is not sufficient evidence to draw any conclusion from this work.

The graphs show that in the approximately thirty test runs, the mean test run was almost identical (a measured difference of 5.3 seconds), marginally shorter in this branch, while the median runtime was measured higher than in `trunk`. Of course, this is just an interesting observation, because again, if there's a real impact here so far, it's indistinguishable from the random variability in the test runs themselves. I'll keep the chart updated as I continue to re-run the performance test suite, and if I can get some E2E test suite data I'll post that as well.
</details>

dmsnell commented:
**Impact on End-to-End Tests Workflow**

![branch-compare-45932-44907](https://user-images.githubusercontent.com/5431237/205760307-fb734ac2-2ba1-4395-a4ec-7cb9ab2c248b.png)

 - this branch, 1868 samples, µ = 15.57 min
 - `trunk`, 608 samples, µ = 17.22 min
 - the data _conclusively suggests this branch is faster for the end-to-end tests workflow_
 - the measured decrease in average runtime is *1m 39s*

Right now we're jumbling all four end-to-end tests into one. This explains the divergence from normality. I'll see if I can separate out the four test jobs since they are independent tests. I don't believe this invalidates the results overall.

<details><summary>Original comment</summary>
With the E2E tests, at around 10 test runs the difference is looking more likely, at a measured decrease in mean test run time of 1 minute, 7 seconds

![branch-compare-45932-44907](https://user-images.githubusercontent.com/5431237/204063291-3c981034-0d71-4f71-b055-5140c778dea5.png)

Something look a little fishy about this data, which may be some oversight on my part when using the scripts I wrote for the Performance workflow with the E2E (which have all four E2E jobs). I'll update of course as I can.
</details>

kevin940726 commented:
@dmsnell Thanks for testing it out! I'm curious how you tested this though. Is there code for the testing script used here?

I think you're missing some important variables in the performance test, we have to take caching into account. If we want to compare it against trunk, we will need to make sure they have the same baseline, do runs with both cache hit and cache miss and compare the results.

In addition, the main motivation of this PR is to increase the performance for running `npm ci` on CI, which isn't directly related to the performance test. Taking the performance test into the equation skews the results and makes it harder to compare, especially when the performance test itself has a very wide range of run time. Since there is no evidence that caching `node_modules` would slow down the performance of the test (we can do a separate test on that to make sure), we should just compare `npm ci` specifically. I think this is why the metrics for E2E tests have a 1-minute improvement while the performance test which is less stable shows less valuable results.

dmsnell commented:
> I think you're missing some important variables in the performance test,

I'm sure I am!

> I'm curious how you tested this though.

What I'm doing is having GitHub re-run the test workflows once they complete. I'm not running the tests locally or in any way that's different than in any typical PR experience. That is, the results I'm measuring are looking at the wall-clock runtime of the CI workflow as reported by GitHub's API which gives the result for each run.

> If we want to compare it against trunk, we will need to make sure they have the same baseline, do runs with both cache hit and cache miss and compare the results.

I'm not sure I agree with this. What I'm trying to measure is the real impact on how long the CI workflows run, and if we make the claim that the caching will help then I think we have to test it against real-world test runs.

There's room to argue that if we merge this than more PRs will start with cached data, but I think we'd still have the possibility that that assumption doesn't hold because among our 1700 branches there are plenty of PRs that won't be forked from the latest `trunk`, and therefore the diversity of merge bases will invalidate the caches in the same way that running the tests in this branch vs. against `trunk` will.

> Since there is no evidence that caching node_modules would slow down the performance of the test (we can do a separate test on that to make sure), we should just compare npm ci specifically.

There are plenty of optimizations that I've investigated in improving the CI workflows that looked promising _in isolation_ but which don't have the same effect when considered inside the system. Some things seem unintuitive, which is why I stress measuring the real results so much.

One possibility here is that GitHub's own `npm` caching is performant enough when it runs in our test containers that it's just as fast (or almost as fast) as having to deal with the cache, or in some cases the cache can cause more harm.

---

My [script](https://gist.github.com/dmsnell/bfcec327f1a8235ba57b9253f218f044) is quite tacky, as I was developing it while exploring the problem and not focusing on building maintainable code. Mostly it amounts to polling GitHub for CI workflow runs and parsing the data; some of what seems unnecessary in the script is the result of a bug in GitHub's API response that I reported and they are working on fixing (when requesting jobs for a given workflow run attempt they only report a link to the jobs for the most-recent run).

I accidentally stopped running the script so I'll pick it up again. Happy to take advice!

kevin940726 commented:
First of all, I don't think the comparison of performance tests makes sense in this context. They are not even doing the same job in trunk and in PRs. In PRs, the performance test is running "Compare performance with trunk" while in trunk it's running "Compare performance with base branch". Comparing them is not an apples-to-apples comparison, and we should at least control our variables to make them as close as possible.

However, I don't think we need to take the full performance test into our equation. We split the job into steps so that we can debug them individually. The steps "Use desired version of NodeJS" and "Npm install" in trunk are responsible for setting up Node.js and installing the dependencies. We replace them with "Setup Node and install dependencies" in this PR. That's the only change in the workflow, anything else doesn't matter. We should tweak our benchmark to only include them (and their cleanup steps if any).

> among our 1700 branches there are plenty of PRs that won't be forked from the latest trunk, and therefore the diversity of merge bases will invalidate the caches in the same way that running the tests in this branch vs. against trunk will.

Yep! This is a valid concern but this is already how it currently works for npm caching in CI. This enhancement won't benefit every PR but it should be beneficial when the cache hits. In the end, that's how caching works, isn't it 😆?

Worth noting that we'll only revalidate the cache when `package-lock.json` is changed, so PRs don't have to be based on the latest trunk. This also follows the same logic of how the official NPM caching works.

> One possibility here is that GitHub's own npm caching is performant enough when it runs in our test containers that it's just as fast (or almost as fast) as having to deal with the cache, or in some cases the cache can cause more harm.

Well, if you look at the installation time reported by GH Actions, you can see that the time reduces significantly when `node_modules` is cached, so this assumption is not valid.

There might be situations where cached `node_modules` has a longer I/O runtime overhead than the installed one, but I highly doubt that 🤔.

Thank you for sharing the script! I'm having trouble understanding some of the contexts though. Which `trunk` are you using when running the test? Are you re-running the same commit on trunk too? Or are you taking the latest 30 commits on trunk. GH Actions by default runs on a "merge commit", so we should use the base commit as "trunk" for the comparison. Note that the base will change if there's a new commit pushed to `trunk` in the meantime.

dmsnell commented:
> First of all, I don't think the comparison of performance tests makes sense in this context.

Maybe there's a misunderstanding. I'm measuring the runtimes of the "End-to-End Tests" and also the "Performance Tests" workflows in CI. I chose these because they are two of the workflows affected by this PR, so I assume if the PR changes them, we should expect an impact there.

I'm not comparing End-to-End tests against Performance tests.

> That's the only change in the workflow, anything else doesn't matter. We should tweak our benchmark to only include them (and their cleanup steps if any).

I'm still challenged by this assertion, as the goal seems to be "Speed up the CI pipeline by caching node_modules." My measurements are looking at the CI pipeline runtime for currently two of the workflows, the End-to-End tests (which take the longest) and the Performance tests (which used to take the longest). We can start measuring any or all of the workflows to see how this affects everything, but the perf tests were convenient since I've been monitoring those for the past month or so, and the end-to-end tests seem relevant since any speedup in the overall process will have to speed those up, or else the improvements won't be felt by anyone (as they will still be waiting for the slowest workflow to complete).

So I don't want to limit ourselves by only looking at a piece in isolation when that piece sits inside a more complicated system where the effects may be different.

> you can see that the time reduces significantly when node_modules is cached, so this assumption is not valid.

This appears to be accurate for the E2E tests, but inaccurate for the performance tests.

> This enhancement won't benefit every PR but it should be beneficial when the cache hits. 

That's absolutely right in theory, and given that we're just sitting here telling GitHub to re-run the CI tests as soon as they finish, we should expect _the best-case of cache hits_ when testing the impact of this PR. If the theory is right then it should be clear from the measurements that this accomplishes its goal.

> it's still taking roughly 2 minutes to install all the packages in every workflow. By also caching node_modules, we can dramatically cut down the installation time by 75%.

According to my latest numbers with 70 test runs, it looks fairly conclusive that it _is_ faster in the end-to-end tests, and the measured difference in average runtime is 1m 37s. After around 30 test runs on the performance tests though, it's not conclusive, but it's looking to move that way in showing that this _is slower in the performance tests_ with a current measured slowdown of adding 47s to the average CI workflow runtime.

I just started re-running the Playwright workflow (last completion took 26min) and the JS Unit Tests workflow (last completion took 4min). It'll be interesting to see what impact this has on those workflows.

> but I highly doubt that 🤔.

All that is to say, given how many false assumptions I found while optimizing the performance tests workflow, I want us to be able to have a credible reason to claim that a change we are making is worth the additional complexity vs. arguing from opinions, guesses, or doubts 😄

It seems clear that this will cut off some time from the E2E workflow, but probably add time to the Performance Tests. When more data comes in it just gives us a better picture to say "I want this tradeoff" or "I don't want this tradeoff."

> Which trunk are you using when running the test? Are you re-running the same commit on trunk too?

I'm using #44907 as a base against which to compare. It's two weeks behind `trunk`, but I'll update it. I became aware of the merge-branch elements when working on the perf tests, but found that practically there it didn't matter much if a little lag came into play as most commits upstream didn't meaningfully impact the tests. We'll find out, as I can easily filter to only use data from the updated `trunk`.

dmsnell commented:
@kevin940726 it seems like many of the workflows are failing due to a `caniuse` issue. I'm not sure if the branch simply needs to be rebased or not, but I hope it's not an issue where caching `npm ci` fails because of non-determinism during `npm ci`

kevin940726 commented:
> Maybe there's a misunderstanding.

Ahh yes definitely. I thought that you were using `trunk`'s workflow which isn't the same as PR's workflow. While in fact, you're using https://github.com/WordPress/gutenberg/pull/44907 as the baseline for "trunk".

> I'm still challenged by this assertion

That's fair! My argument will be that the enhancement in this PR can be viewed in isolation while many of the improvements contributed to the perf test alone cannot be easily split out. I think this PR can still improve the perf test run time, just that we don't have enough data to prove that yet. The run time of the perf test varies a lot and the previous runs might just underperform because of the random factors. We might get a different result if we have more samples. That's just my guess though because the result is counter-intuitive to me and I don't have any explanation for now.

> I want us to be able to have a credible reason to claim that a change we are making is worth the additional complexity vs. arguing from opinions, guesses, or doubts 😄

This is fair too! But I think we should not stop after we create a benchmark, we should try to figure out what happened there. The benchmark is just a tool to help verify that something works. If it creates unexpected results, it's our opportunity to debug that, either in our code (TN) or in the benchmark itself (FN). 

> When more data comes in it just gives us a better picture to say "I want this tradeoff" or "I don't want this tradeoff."

Absolutely! I'm also curious about it and might spend some time creating a benchmark myself too. I'll share more info.

> it seems like many of the workflows are failing due to a caniuse issue. I'm not sure if the branch simply needs to be rebased or not, but I hope it's not an issue where caching npm ci fails because of non-determinism during npm ci

Yeah, that's a very interesting observation! I'll look into that! Thanks! 🙇

kevin940726 commented:
> it seems like many of the workflows are failing due to a `caniuse` issue. I'm not sure if the branch simply needs to be rebased or not

Oh, seems like we just need a rebase after #46093!

dmsnell commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,58 @@
+
+        - name: Post-install
+          run: npm run postinstall
+          shell: bash
```
quick note here: we appear to have prior art in creating a new package for this: see `packages/report-flaky-tests` and `packages/project-management-automation`

I don't have a good feel for preferring one over the other, but it does appear that our existing custom GitHub Actions are created as new packages.

kevin940726 replied:
It's not a new action though, it's just a composite action. Which simply means that it's an abstraction that prevents us from repeating ourselves. We can move it to the `packages` folder but I think that would be overkill at this point.

dmsnell replied:
whatever we call it, I just wanted to note that we have precedence. maybe @youknowriad would have thoughts.

kevin940726 replied:
FWIW, it's what GitHub [calls it](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action).

desrosj replied:
Personally, I am against having any GitHub Action specific code outside of the `.github/workflows` directory. The only times we should make exceptions are when there is a reason to run a script.

A good example is [adding a first time contributor label](https://github.com/WordPress/gutenberg/blob/trunk/packages/project-management-automation/lib/tasks/first-time-contributor-label/index.js). This script would be _very_ awkward to maintain within YAML files.

This change takes advantage of a type of workflow offered in GHA to group a set of shared steps. One or several one line commands grouped together is probably overkill for a new package. So in my opinion, it feels right to include it here rather than within a package (which also has other implications like supporting the package because it was released publicly).

dmsnell commented:
## Impact on Unit Tests workflow

![branch-compare-45932-44907](https://user-images.githubusercontent.com/5431237/205763643-0784cb27-3b29-45b8-8b35-532e5e681123.png)

 - this branch, 1061 samples, µ = 5.18 min
 - trunk, 871 samples, µ = 6.81 min
 - the data _conclusively suggests this branch is faster for the unit test suite_
 - the measured decrease in average runtime is *1m 38s*

dmsnell commented:
@kevin940726 not sure why things changed after the rebase, but now all three of the tests I've been measuring are reporting improvements in this branch.

I'd like to try comparing it against a version without `--ignore-script`. That flag seems at least a little suspect and now I wonder if it was related to the `caniuse` errors. Maybe it's possible that the speedup is partly due to the cache and partly due to the fact that we're skipping steps that we previously relied upon.

Out of curiosity, what led you to add that flag to `npm ci`?

kevin940726 committed:
Try without rebuild and ignore-scripts

kevin940726 committed:
Include npm version to the cache key

kevin940726 committed:
Try reverting the change of graceful-fs

kevin940726 committed:
Update step name

dmsnell commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,50 @@
+        - name: Npm install
+          if: ${{ steps.cache-node_modules.outputs.cache-hit != 'true' }}
+          run: npm ci
+          shell: bash
```
what's the point of `shell: bash` here?

is this the explicit form of…
```
run: |
  npm ci
```

kevin940726 replied:
Nope, it's [required](https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runsstepsshell) for composition action if `run` is specified.

dmsnell replied:
This is interesting to me since it's apparently not required in normal workflow configs, but I confirmed in another branch and everything failed.

kevin940726 commented:
Hi @dmsnell, hope you're doing well! I would love to have some more feedback on this when you have time, no worries if you're not available though. Also c.c. @desrosj since I was suggested that you might be interested in this too 😉 .

dmsnell commented:
@kevin940726 I think one of the best things that can be done for this is simply to monitor it after deploy to measure if it has the expected impact. I expect the impact of the caching to drop once this is in `trunk` and we find a plurality of versions of the dependencies vs. this branch where the dependencies remain mostly unchanged.

how much I expect it to drop I don't know, and I don't know a good way to test this before merging. the more important question is how much we want it to drop in order to balance the costs of increased maintenance, additional debugging when the caching fails us, and the frustration induced when someone is trying to understand why something isn't right.

looks like some recent React-native tests are completing in times similar to the E2E suites. this is good news because as long as those are taking upwards of 30-40 minutes to run, the benefit this caching brings for all the other suites would be locked, but if the longest workflow is taking 20 min right now, then an estimated speedup by around 1 minute and a half could translate into around an 8% speedup, which is pretty good.

we can probably test by iterating over all the open PRs where all workflows have completed, and looking at the one that took the longest to complete. in fact, we may even get this information from the check run or [checks API endpoints](https://docs.github.com/en/rest/checks/runs?apiVersion=2022-11-28#list-check-runs-for-a-git-reference) from Github. I'm still unclear on the terminology even after absorbing myself in it for a while. we should have a reasonable guess within a day or two of data as long as it's mid-week, I would imagine. the determining factor in experiment length is probably how frequently _any_ PR involves changes in the node modules or versions, as that's the thing on which all value in this PR depends. also I think old data sticks around for a month or so? so we're likely limited to a maximum two-week experiment period and that starting immediately after deploy, since after that time we start losing data from the _before_ times.

at some point the benefit this brings won't be worth the development costs, either in a raw reduction of check suite runtime duration or a percentage thereof. I'm curious to know what your target goal is and what you think a reasonable threshold should be for that tradeoff point.

kevin940726 commented:
> I expect the impact of the caching to drop once this is in `trunk` and we find a plurality of versions of the dependencies vs. this branch where the dependencies remain mostly unchanged.

We rarely change the dependencies, at least compared to the number of workflow runs we run. I did a quick search on runs on trunk, out of 642 runs (the workflows that will be impacted by this PR), only 10 of them have changed the `package-lock.json` file. I'm not even counting multiple jobs in a given run (a `end2end-test` run can have 6 jobs for instance). That's about 98% of cache hit. This is only on trunk, I expect the number to be larger if we're also counting PRs.

> the more important question is how much we want it to drop in order to balance the costs of increased maintenance, additional debugging when the caching fails us, and the frustration induced when someone is trying to understand why something isn't right.

What's the cost of maintenance? There's no debugging needed if cache misses for regular contributors, it will simply fallback to what we have now: reinstall everything.

> at some point the benefit this brings won't be worth the development costs, either in a raw reduction of check suite runtime duration or a percentage thereof. I'm curious to know what your target goal is and what you think a reasonable threshold should be for that tradeoff point.

The goal is simple: reduce the time spent on CI, this is only the first step of it. I still fail to understand why you think this is not worth it. If the cache hits, which I expect to be 99% of the cases, then there's going to be roughly 90 to 120 seconds of improvement. If the cache misses, then nothing is changed, it's just going to do whatever we are doing now.

All the numbers collected in this PR show positive results. Unless you have a concrete case against it, I suggest we try this on `trunk` to test it out. We can always revert it if it's not behaving like what we thought anyway 🙂 . WDYT?

dmsnell commented:
> What's the cost of maintenance?

The cost is in understanding and modifying the run scripts.

> There's no debugging needed if cache misses

There are two costs here:
 - if the cache suddenly never hits, then someone will need to figure out why
 - the bigger problem is the cache hitting when it's not supposed to

> why you think this is not worth it

this is definitely _not_ what I said or meant. what I said and meant is that there is a point in all optimizations which come by adding complexity where the benefits and costs balance. I suggested not that this isn't worth it, but that we _measure the impact_ and use that knowledge to confirm it did what we wanted, and if not, that is a clue we can revert it.

why do I keep bringing this up? am I just a fuddy-duddy? 🙃
no, it's because I've found multiple cases where something was _added_ to speed things up and when measured, it either didn't speed things up or it actually slowed things down. this was all done in good faith, but because the impact wasn't measured we were left worse off with more complicated code and nobody realized it.

so the question is a good one, because if you say "90s is a success" and we measure 120s then you know your work made a positive impact. if we measure 10s what do we make of that? or if unintentionally for some quirky reason this slows things down, how do we know?

> I suggest we try this on trunk to test it out

this was literally how I started my last comment, so we are in agreement 😄
I do not believe there is a practical way to measure the full impact of this before merging, or the impact of measuring anything that involves caching at the platform level.

dmsnell reviewed (approved):
this is still marked as a draft, but it's ready for merge.

would still be good to set a target performance impact for this so that after merge we can measure and see if it did what we expected it to.

kevin940726 commented:
Thank you @dmsnell !

> would still be good to set a target performance impact for this so that after merge we can measure and see if it did what we expected it to.

I'd expect all the workflows affected by this PR to get a 1m30s to 2m performance boost if the cache hits. Let's see if that's the case then!

Yeah, I forgot to mark this ready for review 😅. I'll wait for a few other folks to chime in once they get time though!

kevin940726 commented:
Note that the "Report to GitHub" workflow currently fails because it depends on the workflow on `trunk`, which doesn't have this composite action yet. The issue should go away once this is merged.

desrosj commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,50 @@
+
+        - name: Cache node_modules
+          id: cache-node_modules
+          uses: actions/cache@v3
```
Can we ping all 3rd party actions in the workflow to SHA values? This is the only way to pin to an immutable release according to the [GitHub Actions security guide](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions). It's a bit annoying to have to update this, but [Dependabot will call these out](https://github.com/WordPress/gutenberg/pulls?q=is%3Aopen+is%3Apr+author%3Aapp%2Fdependabot+label%3A%22GitHub+Actions%22), and reworking how this works is currently on the GitHub public roadmap (see github/roadmap#592). It looks like some places in the PR have SHAs pinned but others don't.

desrosj commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,50 @@
+          if: ${{ inputs.node-version == '' }}
+          uses: actions/setup-node@v3
+          with:
+              node-version-file: '.nvmrc'
```
The `setup-node` action [documents](https://github.com/actions/setup-node/blob/main/docs/advanced-usage.md#node-version-file) that when both the `node-version` and `node-version-fiile` are present, the `node-version` will be used.

I did some testing, and it seems that when `node-version` is present but no value is supplied, it will fallback correctly to the version defined within the file specified in `node-version-file`. This was true both when an empty string is supplied with `node-version: ''` (see [here](https://github.com/WordPress/wordpress-develop/actions/runs/4387883625/jobs/7683866190#step:3:9)), or just specified as `node-version:` without quotes (see [here](https://github.com/WordPress/wordpress-develop/actions/runs/4387896447/jobs/7683744574#step:3:9)).

With this in mind, we can remove the `if:` conditions and combine both of these steps into one. If a `node-version` is passed to the composite action, great! If not, it will default to the one in `.nvmrc`.

desrosj commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,50 @@
+          uses: actions/cache@v3
+          with:
+              path: '**/node_modules'
+              key: node_modules-${{ runner.os }}-${{ steps.node-npm-version.outputs.NODE_VERSION }}/${{ steps.node-npm-version.outputs.NPM_VERSION }}-${{ hashFiles('package-lock.json') }}
```
I'm thinking that we don't need to include the npm version in the cache key here.

I can't think of a scenario where the version of npm would change independent of the Node.js version.

dmsnell replied:
doesn't this happen regularly in Gutenberg since our `npm` version is still gated at `<7`?

kevin940726 replied:
I don't think we need to include `npm` version here too. Including it will potentially bust the cache more often.

> doesn't this happen regularly in Gutenberg since our `npm` version is still gated at `<7`?

That's because we're using Node 14, which comes with npm v6 by default. npm versions could change independently from Node.js, but they should just be minor releases and backward-compatible.

desrosj replied:
When a new version of Node.js is released, a specific version of npm is bundled with that version. You'll find an [all inclusive manifest here](https://nodejs.org/dist/index.json). As long ass the `npm install -g npm` command is not run, then the version originally bundled with that version of Node.js will be the one installed.

I looked through the code of actions/setup-node. While I wasn't able to confirm this 100%, it seems that new version are pulled and built for use within Action runners using this manifest through the actions/node-versions and actions/versions-package-tools.

Because we only install Node 14.x, the version of npm will always be `< 7`, but the specific version of npm `6.x` installed will be tied to the specific version of Node installed.

dmsnell replied:
Right, but according to the project's `package.json` file we're good to use `node@latest`, which is I think what most people end up doing locally.

https://github.com/WordPress/gutenberg/blob/6434e0bc181bb1c0f5fcf48520be20f1a37afb6c/package.json#L17-L20

so maybe by happenstance we're only using `node@14` in the test suites, but even then, are we guaranteed to run the specific version of 14? or are we pinning it entirely at a specific version?

kevin940726 replied:
> so maybe by happenstance we're only using node@14 in the test suites

`.nvmrc` pins the Node.js version to 14 on CI. FWIW, there's an ongoing PR (#48950) that hopefully can be merged soon to upgrade Node.js and npm to latest on CI.

> but even then, are we guaranteed to run the specific version of 14

Nope, if there's a minor/patch release then we'll automatically use the newer version on CI (basically `^14`). But that's okay since we include the full node version in the cache key.

Note that this PR doesn't affect anyone's local setup whatsoever, they can continue using whatever supported Node.js/npm version they like locally. And the cache only exists on CI, so they won't be affected by it either. This PR only caches `node_modules` on CI, which is a controlled environment.

dmsnell replied:
> if there's a minor/patch release then we'll automatically use the newer version on CI (basically ^14). But that's okay since we include the full node version in the cache key.

I think we're in agreement here. my comment was specifically commenting on how we _do_ expect changes to `node` to occur that don't match `npm`. we're intentionally not using the bundled version of `npm` in the project, but we build the project with changing versions of `node` quite often, since we pin `node` at `>=` a major version.

desrosj commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,50 @@
+name: 'Setup Node'
+description: 'Setup node and install dependencies with caching support'
```
```suggestion
description: 'Setup Node.js and install npm dependencies with caching support'
```

Can we change all occurrences of "NodeJS", "node" etc to "Node.js"? Also, NPM is an acronym, but it's actually `npm` (all [lower case](https://www.npmjs.com/about)). I [recently updated this in WP Core to be consistent](https://core.trac.wordpress.org/ticket/56816).

desrosj commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,50 @@
+description: 'Setup node and install dependencies with caching support'
+inputs:
+    node-version:
+        description: 'Optional node version to pass to setup-node'
```
```suggestion
        description: 'Optional. The Node.js version to use. When not specified, the version specified in .nvmrc will be used.'
```

desrosj commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,50 @@
+              path: '**/node_modules'
+              key: node_modules-${{ runner.os }}-${{ steps.node-npm-version.outputs.NODE_VERSION }}/${{ steps.node-npm-version.outputs.NPM_VERSION }}-${{ hashFiles('package-lock.json') }}
+
+        - name: Npm install
```
```suggestion
        - name: Install npm dependencies
```

desrosj commented on .github/workflows/unit-test.yml:
```diff
@@ -116,10 +110,8 @@ jobs:
-              run: |
-                  npm ci
-                  npm run build
+            - name: Npm build
```
Could we include this in the composite workflow instead of having to include it within each workflow?

We could add a new option for the composite workflow (maybe `run-build`?) with a default of `false` that would run `npm run build` when `true` is passed.

kevin940726 replied:
I don't think this should be added in the composition action. Separating them could help us monitor the time difference between installing the dependencies and building the packages. Also, I think being a bit more explicit helps.

desrosj replied:
That's fair. Building is a separate part of the process from setting up and configuring caching. 👍🏼

desrosj reviewed (changes requested):
This is looking great! Thanks for working on this. I've been meaning to look at this for the Core repo because a few people have mentioned that they still cache their `node_modules` directory in their other projects and it does significantly improve the speed of the related commands.

A bit about caching that I thought would be good to detail. Each repository is allowed 10GB of total cache storage. Once that storage is exceeded, then GHA works to evict caches until bringing the total storage under the limit. When I looked yesterday, the Gutenberg repository was at ~25 GB and GHA had not cleaned up after itself. But looking today, we're at 8.67 GB. The oldest cache was last used 12 hours ago.

I think that we do need to spend some time holistically examining cache key practices to see if we are creating multiple cache entries for the same things anywhere across different workflows. I do think that avoiding any repetition will help us keep things consistent across workflows and potentially help with this.

I made some suggestions and posted a few questions. Would love to give it a second look after these are addressed!

kevin940726 committed:
Code review

kevin940726 commented:
@desrosj Thanks for the reviews! I addressed the feedback 🙇‍♂️

>  When I looked yesterday, the Gutenberg repository was at ~25 GB and GHA had not cleaned up after itself. But looking today, we're at 8.67 GB. The oldest cache was last used 12 hours ago.

Yeah, the caching limits trade-off is mentioned in the PR description too. Note that there are going to be two separate caches while we're still experimenting with this PR though. Caching size will also increase when we're in between updates to the `package-lock.json` file. I think this wouldn't be a problem in the long run, though we should still monitor the cache effect that it may bring.

desrosj commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,42 @@
+          run: npm ci
+          shell: bash
+
+        - name: Post-install
```
Can we add some inline documentation explaining why `postinstall` may need to run run a cache hit occurs?

A few examples in WordPress/wordpress-develop for what I have in mind are [here](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/coding-standards.yml#L90-L95) and [here](https://github.com/WordPress/wordpress-develop/blob/trunk/.github/workflows/phpunit-tests.yml#L115-L126).

desrosj commented on .github/setup-node/action.yml:
```diff
@@ -0,0 +1,42 @@
+          if: ${{ steps.cache-node_modules.outputs.cache-hit == 'true' }}
+          run: |
+              npm run postinstall
+              npx lerna run postinstall
```
What's the difference between running `postinstall` for `npm` and `npx lerna`? Do we need both?

If so, I'm wondering if we should split into two steps. That would make it more clear which of the two fails should either encounter a problem.

kevin940726 replied:
I added some comments explaining this ❤️ .

I don't think we need to split it into two steps though, as they have the same goal.

desrosj commented on .github/workflows/build-plugin-zip.yml:
```diff
@@ -326,7 +326,7 @@ jobs:
                   git config user.email gutenberg@wordpress.org
 
-            - name: Setup Node (for CLI)
+            - name: Setup Node.js (for CLI)
```
```suggestion
            - name: Setup Node.js
```

I don't think we need for CLI here.

desrosj commented on .github/workflows/create-block.yml:
```diff
@@ -20,20 +20,17 @@ jobs:
             fail-fast: false
             matrix:
-                node: [14]
+                node: ['14']
```
Curious why this is needed. Does this cause a failure in the composite workflow because the type is defined as `string`?

kevin940726 replied:
I think either works, but everywhere else uses string. Just want to keep it consistent.

desrosj commented on .github/workflows/publish-npm-packages.yml:
```diff
@@ -49,7 +49,7 @@ jobs:
                   git config user.email gutenberg@wordpress.org
 
-            - name: Setup Node (for CLI)
+            - name: Setup Node.js (for CLI)
```
```suggestion
            - name: Setup Node.js
```

desrosj commented on .github/workflows/pull-request-automation.yml:
```diff
@@ -19,7 +19,7 @@ jobs:
 
-            - name: Use desired version of NodeJS
+            - name: Use desired version of Node.js
               uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3.6.0
```
For this workflow, the `npm install` step is documented inline as "Changing into the action's directory and running `npm install` is much faster than a full project-wide `npm ci`."

With these improvements, I wonder if we can just remove the need for separate cache and install steps within this workflow.

kevin940726 replied:
Yep, but running only on this package is still faster than a full `npm ci`. It's just about 20 seconds faster though. The cache is pretty small so I think it's okay to keep this.

desrosj commented on .github/workflows/unit-test.yml:
```diff
@@ -235,19 +227,14 @@ jobs:
+              uses: ./.github/setup-node
 
-            - name: Npm install and build
+            - name: Npm build
```
Hmm, this seems to be running the Lerna build script. 
```suggestion
            - name: Build Lerna
```

kevin940726 replied:
"Build Lerna" isn't really descriptive either though. I think this should be "Build packages" and just run `npm run build:packages`. But it's already the way it is in trunk, so maybe we should change it in another PR?

desrosj commented on .github/workflows/end2end-test.yml:
```diff
@@ -29,16 +29,11 @@ jobs:
-              run: |
-                  npm ci
-                  npm run build
+            - name: Npm build
```
We're running `npm run build`, but it's not just an npm build. Maybe "Run build scripts" would be more general and accurate. If we change this, it should get changed everywhere.

kevin940726 replied:
Yeah, let's do this in another PR though.

desrosj reviewed (commented):
Looking good! Here's a few more things that popped out on re-review.

kevin940726 committed:
Add some comments

desrosj reviewed (approved):
Thanks @kevin940726 I'm OK with merging this to try it out!

kevin940726 commented:
Thanks! Let's try it and revert it if anything goes wrong 😅 !

desrosj commented:
@kevin940726 I know it's only been ~4 days, but wanted to circle back to see if you'd observed any impacts of this change, either positive or negative.

desrosj commented:
One thing I noticed that's a little annoying is that there is an annotation added when both `node-version` and `node-version-file` are present.

![Screenshot 2023-03-20 at 13 39 12](https://user-images.githubusercontent.com/359867/226441855-96fb52ed-719f-41de-876f-1c665cc3148a.png)

kevin940726 commented:
Sure! I did some basic analysis on the ["Unit Tests"](https://github.com/WordPress/gutenberg/actions/workflows/unit-test.yml?query=is%3Asuccess+branch%3Atrunk) workflow before and after the commit on trunk.

I took 30 runs each and here are the results:

Before:
> Mean: 10m 18s
> Median: 10m 21s
> Standard deviation: 42s
 
After:
> Mean: 8m 51s
> Median: 8m 29s
> Standard deviation: 1m 17s

Will be interesting to do more analysis on other workflows as well, but I don't have the time rn 😅 .

> One thing I noticed that's a little annoying is that there is an annotation added when both `node-version` and `node-version-file` are present.

Yeah, turns out we do want the `if` statement afterall 😅 .
